subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -405,18 +453,36 @@ public Configuration resolutionStrategy(Closure closure) {
     }
 
     private void validateMutation(boolean lenient) {
-        boolean userAlreadyNagged = false;
-        if (getState() != State.UNRESOLVED) {
-            if (!lenient) {
-                throw new InvalidUserDataException(String.format("Cannot change %s after it has been resolved.", getDisplayName()));
-            } else {
-                userAlreadyNagged = true;
+        switch (state) {
+            case UNOBSERVED:
+                // Nothing from the configuration has been observed yet, can change anything.
+                break;
+            case OBSERVED:
+                // The configuration has been used in a resolution, and it is deprecated for
+                // build logic to change any dependencies, artifacts, exclude rules or parent
+                // configurations (non-lenient properties).
+                if (!lenient) {
+                    DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to change %s after it has been included in dependency resolution", getDisplayName()));
+                }
+                break;
+            case TASK_DEPENDENCIES_RESOLVED:
+                // The task dependencies for the configuration have been calculated using
+                // Configuration.getBuildDependencies(). It is deprecated for build logic to
+                // change anything about the configuration.
                 DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to change %s after it has been resolved", getDisplayName()));
```

Comment:
Would be good to describe that it "has been resolved to determine task dependencies".
Should mention that the change will force a re-resolve of the configuration when it is used.